main.go

```diff
@@ -178,24 +179,158 @@ func isPregnant(kittyID uint64) (bool, error) {
+	if int(matronID) >= len(kitties) {
+		return false, fmt.Errorf("No kitty with this ID %d available.", matronID)
+	}
+
+	if int(sireID) >= len(kitties) {
+		return false, fmt.Errorf("No kitty with this ID %d available.", sireID)
+	}
+
+	if sireID == matronID {
+		return false, nil
+	}
+
+	matron := kitties[matronID]
+	sire := kitties[sireID]
+
+	if matron.MatronID == sireID || matron.SireID == sireID {
+		return false, nil
+	}
+	if sire.MatronID == sireID || sire.SireID == sireID {
+		return false, nil
+	}
+
+	if sire.Generation == 0 || matron.Generation == 0 {
+		return true, nil
+	}
+
+	if sire.MatronID == matron.MatronID || sire.MatronID == matron.SireID {
+		return false, nil
+	}
+	if sire.SireID == matron.MatronID || sire.SireID == matron.SireID {
+		return false, nil
+	}
+
 	return true, nil
 }
 
-func (c *KittyContract) CanBreedWith(ctx contractapi.TransactionContextInterface, sire uint64, matron uint64) (bool, error) {
+func (c *KittyContract) CanBreedWith(ctx contractapi.TransactionContextInterface, sireID, matronID uint64) (bool, error) {
+	if int(matronID) >= len(kitties) || matronID == 0 {
+		return false, fmt.Errorf("No matron with this ID %d available.", matronID)
+	}
+
+	if int(sireID) >= len(kitties) || sireID == 0 {
+		return false, fmt.Errorf("No sire with this ID %d available.", sireID)
+	}
+
+	ok, err := isValidMatingPair(matronID, sireID)
+	if err != nil || !ok {
+		return false, err
+	}
+
+	ok, err = isSiringPermitted(matronID, sireID)
+	if err != nil || !ok {
+		return false, err
+	}
+
 	return true, nil
 }
 
-func breedWith(sire uint64, matron uint64) error {
+func breedWith(sireID, matronID uint64) error {
+	if int(matronID) >= len(kitties) {
+		return fmt.Errorf("No kitty with this ID %d available.", matronID)
+	}
+
+	if int(sireID) >= len(kitties) {
+		return fmt.Errorf("No kitty with this ID %d available.", sireID)
+	}
+
+	matron := kitties[matronID]
+
+	matron.SiringWithID = sireID
+	triggerCooldown(matronID)
+	triggerCooldown(sireID)
+
+	sireAllowedToAddress[sireID] = ""
+	sireAllowedToAddress[matronID] = ""
+
+	owner := kittyIndexToOwner[matronID]
+
+	g_event["Pregnant"] = map[string]interface{}{"owner": owner, "matronID": matronID, "sireID": sireID, "matronCooldown": matron.CooldownEnd}
+
 	return nil
 }
 
-func (c *KittyContract) BreedWithAuto(ctx contractapi.TransactionContextInterface, sireID uint64, matronID uint64) error {
-	return nil
+func (c *KittyContract) BreedWithAuto(ctx contractapi.TransactionContextInterface, sireID, matronID uint64) error {
+	if int(matronID) >= len(kitties) {
+		return fmt.Errorf("No kitty with this ID %d available.", matronID)
+	}
+
+	if int(sireID) >= len(kitties) {
+		return fmt.Errorf("No kitty with this ID %d available.", sireID)
+	}
+
+	clientID, err := ctx.GetClientIdentity().GetID()
+	if err != nil {
+		return err
+	}
+
+	matronOwner := kittyIndexToOwner[matronID]
+
+	if matronOwner != clientID {
+		return fmt.Errorf("Caller must be the owner of the matron kitty.")
+	}
+
```

Comment:
You have checked at this point that the matron is owned by the message sender. After that, in the solidity implementation, there is a requirement for isSiringPermitted, eg:
```solidity
require(_isSiringPermitted(_sireId, _matronId));
```
This is after the owner is checked and before the ReadyToBreed functions are called. I think this isSiringPermitted() is missing here.